README.md
# Vault kv migration tool

**Author: Mick Miller**

**Date: 2020-12-16**

 

This bash script was built primarily to migrate a ton of secrets from an old, open-source version of Vault to Vault Enterprise. Surprisingly, there was no migration tool available from Hashicorp at the time of this work, so I created one.

## Contents

1. The config.json file
2. Installation instructions
3. Running the script
4. Useful commands and random-ish stuff
5. References
6. Acknowledgements

A couple of notes before diving in: 

* Back up your source secrets engine(s). 
* Understand what this script is doing before you run it and tweak as needed.
* You will need to configure the config.json file for your specific use case.

> NOTE: You may notice the pattern `# echo "DEBUG ${LINENO}: "Some string"` in the script. This is used for debugging the script. I left them in for you in case you wanted to trace the code; sorry if it irritates you.

---

## 1. The config.json file

This configuration file is used to reduce the amount of command line arguments and limit the arguments to:

* The path to find the secrets; and
* Not storing the tokens in the configurations or code.

config.json

```
{ 
    "type_val": "general",
    "src_url": "https://old_vault.example.com",
    "dest_url": "https://new_vault.example.com",
    "tmp_file": "./tmp.json",
    "config_file":"./config.yaml"
}
```

### Explanation of keys and values

| Key           | Value                                                                           |
| ---           | -----                                                                           |
| `type_value`  | The type of secrets engine in the source vault instance                         |
| `src_url`     | The source vault instance URL                                                   |
| `dest_url`    | the destination vault instance URL                                              |
| `tmp_file`    | The name of the output temp JSON file; you should not need to change this value |
| `config_file` | The name of this file                                                           |

---

## 2. Installation instructions
=======

The code assumes that both the Hashi Vault client and jq are installed before you start and tests for the presence of both.

### Installing Vault and jq

If you are using the Homebrew package manager on mac OS, run the following:

```
 $ brew install jq
 $ brew install vault
```

This script has not been tested on Windows or Linux, only macOS. I will test Ubuntu at some point and refactor as needed.

### Installing the script

```
# Clone the repo and then change to the directory.
$ git clone <this repo url>
$ cd vault_kv_migration

# Run the script
$ vault_kv_migration.sh -s "${SRC_TOKEN}" -d "${DEST_TOKEN}" -p "${VAULT_PATH}"
Usage:
  Format : ./vault_kv_migrator.sh {source token} {destination token} {path}
  Example: ./vault_kv_migrator.sh -s xxxxxxx -d xxxxxxx -p /secret/cnn/
  Note   : A trailing slash in path is required."
```

---

## 3. Running the script

Command line arguments description

```
$ vault_kv_migration.sh -s "${SRC_TOKEN}" -d "${DEST_TOKEN}" -p "${VAULT_PATH}"

Usage:
  Format : ./vault_kv_migrator.sh {source token} {destination token} {path}
  Example: ./vault_kv_migrator.sh -s xxxxxxx -d xxxxxxx -p /secret/cnn/
  
  Note   : A trailing slash in path is required."
```  

---

## 4. Useful commands and random-ish stuff

## Log in to Vault

```
export VAULT_ADDR=https://vault.example.com
vault login -method=ldap mount=ad username=mick
```

View login name**
```
TOKEN=$(vault print token) | vault token lookup $TOKEN
```

### Log out of Vault

I know; strange, but effective.

```
rm ~/.vault-token
```

### Check status

```
vault status
```

### Create a secretes engine

```
vault secrets enable -path=kv kv-v2
vault kv enable-versioning kv
vault write kv/config max_versions=4
```

### Create Active Directory groups

```
vault write auth/ad/groups/admin-group policies=admins,app1,app1
vault write auth/ad/groups/app-group policies=app1,app2
vault read auth/ad/groups/admin-group
vault read auth/ad/groups/app-group
vault list auth/ad/groups
```

### Create a policy

```
vault policy write admins ./admins-policy.hcl
vault policy list
vault policy read admins
```

### Create two secrets

```
vault kv put kv/anthos/test test=12345
vault kv put kv/gcp/test test=12345

vault kv list kv/anthos
vault kv list kv/gco
vault kv list -format=json kv/gcp
```

---

## 5. References

Below are a couple of good references for learning the Terraform Vault provider information:

* https://learn.hashicorp.com/tutorials/vault/codify-mgmt-enterprise
* https://registry.terraform.io/providers/hashicorp/vault/latest/docs

---

## 6. Acknowledgements

Many thanks to the following folks:

* agaudreault-jive (https://github.com/hashicorp/vault/issues/5275)
* user2599522 (https://stackoverflow.com/a/61000422)
* kir4h (https://github.com/kir4h/rvault)


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [0.1.0] - 2020-12-17

First release

vault_kv_migrator.sh
#!/bin/bash
#====================================================================================================================
# Name:         vault_kv_migrator.sh
# Created by:   Mick Miller
# Created on:   2020-12-15
# Updated on:   check github commits
# References: 
#   https://github.com/hashicorp/vault/issues/5275
#   https://stackoverflow.com/questions/60932147/hashicorp-vault-export-key-from-one-vault-import-into-another-vault
# ===================================================================================================================

# TODO:
# summary page at the end with stats
# - For example: time to run, number of directories and keys, etc.


# set -e
clear

# Get command line args and put them in an array
ARGS=( "$@" )
NUM_ARGS=( "$#" )

# Get tokens and path from args
SRC_TOKEN="${ARGS[1]}"
DEST_TOKEN="${ARGS[3]}"
VAULT_PATH="${ARGS[5]}"

# Run time parameters
CONFIG_FILE="./config_migrate.yaml"

# Command line arg missing
function cmd_line_error {
  arg1="${1}"
  echo
  echo ===============================================================================
  echo "Command line arguments descirption."
  echo "You typed:" "${0}" -s "${SRC_TOKEN}" -d "${DEST_TOKEN}" -p "${VAULT_PATH}"
  echo
  echo "Usage: "
  echo "format : " "${0} " -s {source token} -d {destination token} -p {path}
  echo "example: " "${0} " "-s xxxxxxx -d xxxxxxx -p /secret/cnn/"
  echo "Note: trailing slash in path is important"
  echo "$arg1"
  echo ===============================================================================
}

# Start of job logging
function start_job {
  arg1="${1}"
  echo
  echo ===============================================
  echo Start "${arg1}....."
  echo
}

# End of job logging
function end_job {
  arg1="${1}"
  echo
  echo Done "${arg1}"
  echo ===============================================
  echo
}

# Test for vault installed
function test_for_vault {
  ver_vault=$(vault version)
  if [[ ${?} -ne 0 ]]; then
    cmd_line_erros "Error - Can't find a version of vault in the current path"
    exit
  fi
}

# Test for jq installed
function test_for_jq {
  ver_jq=$(jq --version)
  if [[ $? -ne 0 ]]; then
    cmd_line_erros "Error - Can't find a version of jq in the current path"
    exit
  fi
}

# Read command line variables
function test_cmd_line_args {
  if [[ "${NUM_ARGS}" -ne 6 ]]; then
    cmd_line_error "Error - Incorrect command line arguments"
    exit
  fi
}

# Check for trailing slash in path
function test_for_trailing_slash {
  if [[ "${VAULT_PATH}" != */ ]]; then 
    cmd_line_error "Error - Missing trailing slash in path" 
    exit
  fi
}

# Load configuration yaml file
function load_config {
  if [[ ! -f "${CONFIG_FILE}" ]]; then 
    cmd_line_error "Error - Missing ${CONFIG_FILE} in current path" 
    exit
  else
    # load the file variables
    TMP_FILE=$(cat ${CONFIG_FILE} | jq .tmp_file)
    TMP_FILE=$(sed -e 's/^"//' -e 's/"$//' <<<$TMP_FILE) # remove double quotes
    TYPE_VAL=$(cat ${CONFIG_FILE} | jq .type_val)
    SRC_URL=$(cat ${CONFIG_FILE} | jq .src_url)
    SRC_URL=$(sed -e 's/^"//' -e 's/"$//' <<<$SRC_URL) # remove double quotes
    DEST_URL=$(cat ${CONFIG_FILE} | jq .dest_url)   
    DEST_URL=$(sed -e 's/^"//' -e 's/"$//' <<<$DEST_URL) # remove double quotes
  fi
}

# Get all list of all key paths then recuse path for all keys values.
function get_keys_from_dev {

  if [[ "$VAULT_PATH" ]]; then
    # Make sure the path always end with '/'
    VAULTS=("${VAULT_PATH%"/"}/")
    # # echo "DEBUG ${LINENO}: Using this path: $VAULT_PATH"
  else
    # Get a list of all secrets engines of a specific type if no path provided.  TODO - Not sure I need this test
    VAULTS=$(vault secrets list -format=json | jq -r 'to_entries[] | select(.value.type == "$TYPE_VAL") | .key')
    # echo "DEBUG ${LINENO}: Found these vaults $VAULTS"
  fi
  
  # Might not need a for loop here because I am passing in the path
  for VAULT in "${VAULTS}"; do
    # echo "DEBUG ${LINENO}: Main for loop control given path: " $VAULT "in" $VAULTS
    recurse "${VAULT}"
  done
}

# Recursive function that will
# - List all the secrets in the given $PATH
# - Call itself for all path values in the given $PATH
function recurse {
    arg1="$1"
    # # echo "DEBUG ${LINENO}: Recuse with agrument: " $arg1
    
    # Set to source vault
    export VAULT_TOKEN="$SRC_TOKEN"
    export VAULT_ADDR=$SRC_URL
    # echo "DEBUG ${LINENO}: Using Vault Token: $VAULT_TOKEN and Vault URL: $VAULT_ADDR"
    local readonly KEY_PATH="${arg1}"

    # List all folders for a given path 
    LIST_DIRS_KEYS=$(vault kv list -format=json "${KEY_PATH}" 2>&1)
    # echo "DEBUG ${LINENO}: Keys under: $arg1" "$LIST_DIRS_KEYS"
    status="${?}"
    if [ ! "${status}" -eq 0 ];
    then
        if [[ "${LIST_DIRS_KEYS}" =~ "permission denied" ]]; then
            return
        fi
        >&2 # echo "DEBUG ${LINENO}: ${LIST_DIRS_KEYS}"
    fi

    # Read the keys in json format 
    for SECRET in $(echo "${LIST_DIRS_KEYS}" | jq -r '.[]'); do
        # echo "DEBUG ${LINENO}: For loop secret:  $SECRET in $LIST_DIRS_KEYS"
        if [[ "${SECRET}" == */ ]]; then # found a parent folder, recurse the folder for keys
            # echo "DEBUG ${LINENO}: Found parent folder at $KEY_PATH$SECRET"
            recurse "${KEY_PATH}${SECRET}"
        else # no trailing slash means this is a key and not a folder or subfolder
            echo =====================================================================================================
            # echo "DEBUG ${LINENO}: Found key at $KEY_PATH$SECRET"
            JSON_KEY=$(vault kv get -format=json -field data ${KEY_PATH}${SECRET})
            # echo "DEBUG ${LINENO}: Found key values $JSON_KEY" " - call write_key_to_prod"
            write_key_to_prod "${JSON_KEY}" "${KEY_PATH}${SECRET}"
        fi
    done
}

function write_key_to_prod {
  JSON_KEY2="${1}"
  KEY_PATH2="${2}"

  # production vault
  export VAULT_TOKEN="${DEST_TOKEN}"
  export VAULT_ADDR="${DEST_URL}"
  
  # # echo "DEBUG ${LINENO}: Destination token: "
  # TOKEN=$(vault print token) | vault token lookup -format=json $TOKEN | jq .data.id
  
  echo "Get key from DV path: " "${KEY_PATH2}${SECRET2}"
  DST_PATH="${KEY_PATH2/secret/kv}"
  echo "Write key to destination path: " "${DST_PATH}"   
  # echo "DEBUG ${LINENO}: Put key values: " $JSON_KEY2
  # echo "DEBUG ${LINENO}: Using json: ${JSON_KEY2} to put in file: ${TMP_FILE}" 
  echo "${JSON_KEY2}" > ${TMP_FILE} 2>&1
  
  echo -----------------------------------------------------------------------------------------------------
  echo "Creating new key in vault PR ${DST_PATH} with token: $DEST_TOKEN at $DEST_URL"
  PUT_RESULT=$(vault kv put "${DST_PATH}" @"${TMP_FILE}" 2>&1)
  echo $PUT_RESULT
  
  # GET_RESULT=$(vault kv get $DST_PATH)
  # echo "DEBUG ${LINENO}: Current KV = " $GET_RESULT 
  echo =====================================================================================================
  echo
  
  # set back to dev vault
  export VAULT_TOKEN="${SRC_TOKEN}"
  export VAULT_ADDR="${SRC_URL}"
  
  # echo ""DEBUG ${LINENO}:Source token: " 
  # TOKEN=$(vault print token) | vault token lookup -format=json $TOKEN | jq .data.id
  echo "Start reading new key..."
}

function list_new_keys {
  arg1=$1
  echo =====================================================================================================
  echo "List vault keys for input path: ${arg1}"
  echo
  vault kv list "${arg1}"
  echo =====================================================================================================
}

function clean_up_tmp_file {
  result=$(rm -fv ./tmp.json)
  echo "Cleaning up tmp file: " $result
}

# ========================
# main
# ========================
# Pre-flight tests
test_for_vault
test_for_jq
test_cmd_line_args
test_for_trailing_slash

# Load the configs files and initialize variables
load_config

# Iterate on all kv engines or start from the path provided by the user
start_job " Hashi Vault migration..."

get_keys_from_dev
list_new_keys "${VAULT_PATH}"
clean_up_tmp_file

end_job " Hashi Vault migration."


dv_pr_vault.sh
#!/bin/bash
#====================================================================================================================
# Name:         dv_pr_vault.sh
# Created by:   Mick Miller
# Created on:   2020-12-15
# References: 
#   https://github.com/hashicorp/vault/issues/5275
#   https://stackoverflow.com/questions/60932147/hashicorp-vault-export-key-from-one-vault-import-into-another-vault
# ===================================================================================================================

# TODO:

# set -e
clear

# Get command line args and put them in an array
ARGS=( "$@" )
NUM_ARGS=( "$#" )

# Get tokens and path from args
SRC_TOKEN="${ARGS[1]}"
DEST_TOKEN="${ARGS[3]}"
VAULT_PATH="${ARGS[5]}"

# Run time parameters
CONFIG_FILE="./config.yaml"

# Command line arg missing
function cmd_line_error {
  arg1="${1}"
  echo
  echo ===============================================================================
  echo "Missing command line values."
  echo "You typed:" "${0}" -s "${SRC_TOKEN}" -d "${DEST_TOKEN}" -p "${VAULT_PATH}"
  echo
  echo "Usage: "
  echo "format : " "${0} " {source token} {destination token} {path}
  echo "example: " "${0} " "-s xxxxxxx -d xxxxxxx -p /secret/cnn/"
  echo "Note: trailing slash in path is important"
  echo "$arg1"
  echo ===============================================================================
}

# Start of job logging
function start_job {
  arg1="${1}"
  echo
  echo ===============================================
  echo Start "${arg1}....."
  echo
}

# End of job logging
function end_job {
  arg1="${1}"
  echo
  echo Done "${arg1}"
  echo ===============================================
  echo
}

# Test for vault installed
function test_for_vault {
  ver_vault=$(vault version)
  if [[ ${?} -ne 0 ]]; then
    cmd_line_erros "Error - Can't find a version of vault in the current path"
    exit
  fi
}

# Test for jq installed
function test_for_jq {
  ver_jq=$(jq --version)
  if [[ $? -ne 0 ]]; then
    cmd_line_erros "Error - Can't find a version of jq in the current path"
    exit
  fi
}

# Read command line variables
function test_cmd_line_args {
  if [[ "${NUM_ARGS}" -ne 6 ]]; then
    cmd_line_error "Error - Incorrect command line arguments"
    exit
  fi
}

# Check for trailing slash in path
function test_for_trailing_slash {
  if [[ "${VAULT_PATH}" != */ ]]; then 
    cmd_line_error "Error - Missing trailing slash in path" 
    exit
  fi
}

# Load configuration yaml file
function load_config {
  if [[ -f "${CONFIG_FILE}" ]]; then 
    cmd_line_error "Error - Missing ${CONFIG_FILE} in current path" 
    exit
  else
    # load the file variables
    TMP_FILE=$(cat ./config.json | jq .tmp_file)
    TYPE_VAL=$(cat ./config.json | jq .type_val)
    DV_URL=$(cat ./config.json | jq .dv_url)
    DV_URL=$(sed -e 's/^"//' -e 's/"$//' <<<$DV_URL) # remove double quotes
    PV_URL=$(cat ./config.json | jq .dv_url)   
    PV_URL=$(sed -e 's/^"//' -e 's/"$//' <<<$PV_URL) # remove double quotes
    # Run time parameters
  fi
}


# Get all list of all key paths then recuse path for all keys values.
function get_keys_from_dev {

  if [[ "$VAULT_PATH" ]]; then
    # Make sure the path always end with '/'
    VAULTS=("${VAULT_PATH%"/"}/")
    # echo "DEBUG ${LINENO}: Using this path: $VAULT_PATH"
  else
    # Get a list of all secrets engines of a specific type if no path provided.  TODO - Not sure I need this test
    VAULTS=$(vault secrets list -format=json | jq -r 'to_entries[] | select(.value.type == "$TYPE_VAL") | .key')
    # echo "DEBUG ${LINENO}: Found these vaults $VAULTS"
  fi
  
  # Might not need a for loop here because I am passing in the path
  for VAULT in "${VAULTS}"; do
    echo "DEBUG ${LINENO}: Main for loop control given path: " $VAULT "in" $VAULTS
    recurse "${VAULT}"
  done
}

# Recursive function that will
# - List all the secrets in the given $PATH
# - Call itself for all path values in the given $PATH
function recurse {
    arg1="$1"
    # echo "DEBUG ${LINENO}: Recuse with agrument: " $arg1
    # Set to source vault
    export VAULT_TOKEN="$SRC_TOKEN"
    # echo "DEBUG ${LINENO}: SRV_TOKEN = $SRC_TOKEN"
    export VAULT_ADDR=$DV_URL
    # echo "DEBUG ${LINENO}: VAULT_ADDR = $VAULT_ADDR"
    local readonly KEY_PATH="${arg1}"

    # List all folders for a given path 
    LIST_DIRS_KEYS=$(vault kv list -format=json "${KEY_PATH}" 2>&1)
    echo "DEBUG ${LINENO}: Keys under: $arg1" "$LIST_DIRS_KEYS"
    status="${?}"
    if [ ! "${status}" -eq 0 ];
    then
        if [[ "${LIST_DIRS_KEYS}" =~ "permission denied" ]]; then
            return
        fi
        >&2 echo "DEBUG ${LINENO}: ${LIST_DIRS_KEYS}"
    fi

    # Read the keys in json format 
    for SECRET in $(echo "${LIST_DIRS_KEYS}" | jq -r '.[]'); do
        # echo "DEBUG ${LINENO}: For loop secret:  $SECRET in $LIST_DIRS_KEYS"
        if [[ "${SECRET}" == */ ]]; then # found a parent folder, recurse the folder for keys
            # echo "DEBUG ${LINENO}: Found parent folder at $arg1$SECRET"
            recurse "${KEY_PATH}${SECRET}"
        else # no trailing slash means this is a key and not a folder or subfolder
            echo =====================================================================================================
            # echo "DEBUG ${LINENO}: Found key at $KEY_PATH$SECRET"
            JSON_KEY=$(vault kv get -format=json -field data ${KEY_PATH}${SECRET})
            # echo "DEBUG ${LINENO}: Found key values $JSON_KEY" " - call write_key_to_prod"
            write_key_to_prod "${JSON_KEY}" "${KEY_PATH}${SECRET}"
        fi
    done
}

function write_key_to_prod {
  JSON_KEY2="${1}"
  KEY_PATH2="${2}"

  # production vault
  export VAULT_TOKEN="${DEST_TOKEN}"
  export VAULT_ADDR="${PR_URL}"
  
  # echo "Destination token: "
  # TOKEN=$(vault print token) | vault token lookup -format=json $TOKEN | jq .data.id
  
  echo "Get key from DV path: " "${KEY_PATH2}${SECRET2}"
  PR_PATH="${KEY_PATH2/secret/kv}"
  echo "Write key to PR path: " "${PR_PATH}"   
  # echo "DEBUG ${LINENO}: Put key values      : " $JSON_KEY2
  echo "${JSON_KEY2}" > "${TMP_FILE}" 2>&1

  echo -----------------------------------------------------------------------------------------------------
  echo "Creating new key in vault PR ${PR_PATH}"

  PUT_RESULT=$(vault kv put "${PR_PATH}" @"${TMP_FILE}" 2>&1)
  # echo $PUT_RESULT
  
  # GET_RESULT=$(vault kv get $PR_PATH)
  # echo $GET_RESULT
  echo =====================================================================================================
  echo
  
  # set back to dev vault
  export VAULT_TOKEN="${SRC_TOKEN}"
  export VAULT_ADDR="${DV_URL}"
  
  # echo "Source token: " 
  # TOKEN=$(vault print token) | vault token lookup -format=json $TOKEN | jq .data.id
  echo "Start reading new key..."
}

function list_new_keys {
  arg1=$1
  echo =====================================================================================================
  echo "List vault keys for input path: ${arg1}"
  echo
  vault kv list "${arg1}"
  echo =====================================================================================================
}

function clean_up_tmp_file {
  result=$(rm -fv ./tmp.json)
  echo "Cleaning up tmp file: " $result
}

# ========================
# main
# ========================
# test command line args
test_for_vault
test_for_jq
test_cmd_line_args
test_for_trailing_slash
load_config

# Iterate on all kv engines or start from the path provided by the user
start_job " Hashi Vault migration..."
get_keys_from_dev
list_new_keys "${VAULT_PATH}"
clean_up_tmp_file
end_job " Hashi Vault migration."